ember.md

```diff
@@ -0,0 +1,461 @@
+informative error when called with `null` or `undefined`.
+
+Although when defining a method in a controller, component, etc. you
+can be fairly sure `this` is an Ember Object, for consistency with the
+above, we still use `get`/`set`.
+
+```js
+// Good
+import Ember from 'ember';
+
+const { get, set } = Ember;
+
+set(object, 'isSelected', true);
+get(object, 'isSelected');
+get(this, 'isPending');
+
+// Ok
+this.get('isPending');
+
+// Bad
+object.set('isSelected', true);
+object.get('isSelected');
+```
+
+### Use brace expansion
+
+This allows much less redundancy and is easier to read.
+
+Note that **the dependent keys must be together (without space)** for the brace expansion to work.
+
+```js
+// Good
+fullName: computed('user.{firstName,lastName}', function _fullName{
+  // Code
+}),
+
+// Bad
+fullName: computed('user.firstName', 'user.lastName', function _fullName{
+  // Code
+}),
+```
+
+### Trailing Commas
+
+Trailing commas must be used for multiline literals, and must not be used for single-line literals.
+
+```js
+// good
+const { get, computed } = Ember;
+const {
+  get,
+  computed,
+} = Ember;
+
+export default Ember.Object.extend({
+  foo() {
+  },
+
+  actions: {
+    bar() {
+    },
+  },
+});
+```
+
+*Trailing commas are easier to maintain; one can add or remove properties without having to maintain the commas. Also produces cleaner git diffs.*
+
+### All functions must be named
+
+Any function declared with the `function` keyword must have a name. Functions declared with the es6 short "fat arrow" syntax do not need a name.
+
+This helps with debugging, as the name of the function shows up in the backtrace instead of `(anonymous function)`.
+
+When the function is used as part of a macro to define a property, we have a convention of using the property name prefixed with an underscore.
+
+```js
+// good
+
+things.map((thing) => get(thing, 'name'));
+
+fullName: computed('firstName', 'lastName', function _fullName() {
+  // Code
+}),
+
+// bad
+
+fullName: computed('firstName', 'lastName', function() {
+  // Code
+}),
+```
+
+## Organizing your modules
+
+Ordering a module's properties in a predictable manner will make it easier to
+scan.
+
+1. __Plain properties__
+
+   Start with properties that configure the module's behavior. Examples are
+   `tagName` and `classNames` on components and `queryParams` on controllers and
+   routes. Followed by any other simple properties, like default values for properties.
+
+2. __Single line computed property macros__
+
+   E.g. `alias`, `sort` and other macros. Start with service injections. If the
+   module is a model, then `attr` properties should be first, followed by
+   `belongsTo` and `hasMany`.
+
+3. __Multi line computed property functions__
+
+4. __Lifecycle hooks__
+
+   The hooks should be chronologically ordered by the order they are invoked in.
+
+5. __Functions__
+
+   Public functions first, internal functions after.
+
+6. __Actions__
```

Comment:
Oh huh. There is a weird distinction here. I think that internal functions that are actually part of the defined class should go above actions. However, sometimes we define utility functions outside of the class. I'm not sure about the terminology here, but I'll just pick some. For example:

```js
export default Ember.Object.extend({
  someFunction() {
    // this is a public function
  },

  _someOtherFunction() {
    // this is a private (or internal function) that is part of the class
  },

  actions: {
  },
});

function _someOtherFunction() {
  // This is a utility function that has to be after actions because it is actually outside of the class.
};
```